src/backend/native_gpu_future.rs

```diff
@@ -1,72 +1,121 @@
 use std::future::Future;
 use std::pin::Pin;
-use std::sync::{Arc, Mutex};
+use std::sync::Arc;
 use std::task::{Context, Poll, Waker};
+use parking_lot::Mutex;
+use crate::BufferAddress;
 
-struct GpuFutureInner<T> {
-    id: wgc::id::DeviceId,
-    result: Option<T>,
-    waker: Option<Waker>,
+enum WakerOrResult<T> {
+    Waker(Waker),
+    Result(T),
 }
 
 /// A Future that can poll the wgpu::Device
 pub struct GpuFuture<T> {
-    inner: Arc<Mutex<GpuFutureInner<T>>>,
+    data: Arc<Data<T>>,
+}
+
+pub enum OpaqueData {}
+
+struct Data<T> {
+    device_id: wgc::id::DeviceId,
+    buffer_id: wgc::id::BufferId,
+    size: BufferAddress,
+    waker_or_result: Mutex<Option<WakerOrResult<T>>>,
 }
 
 /// A completion handle to set the result on a GpuFuture
 pub struct GpuFutureCompletion<T> {
-    inner: Arc<Mutex<GpuFutureInner<T>>>,
+    data: Arc<Data<T>>,
 }
 
 impl<T> Future for GpuFuture<T>
 {
     type Output = T;
 
     fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
-        // grab a clone of the Arc
-        let arc = Arc::clone(&self.get_mut().inner);
+        let data = &self.get_mut().data;
 
         // grab the device id and set the waker, but release the lock, so that the native callback can write to it
         let device_id = {
-            let mut inner = arc.lock().unwrap();
-            inner.waker.replace(context.waker().clone());
-            inner.id
+            let mut waker_or_result = data.waker_or_result.lock();
+
+            if let Some(WakerOrResult::Result(res)) = waker_or_result.take() {
+                return Poll::Ready(res);
+            }
+
+            waker_or_result.replace(WakerOrResult::Waker(context.waker().clone()));
+
+            data.device_id
         };
 
         // polling the device should trigger the callback
-        wgn::wgpu_device_poll(device_id, true);
+        wgn::wgpu_device_poll(device_id, false);
```

Comment:
I've been thinking about this a lot lately and these are some of the conclusions I've been arriving at:

- `Future::poll` should not call `wgpu_device_poll` at all.

- `wgpu_device_poll` should not fire **the user provided callback** (at least in the scope of `wgpu-rs`). 

- `wgpu_device_poll` should execute some kind of shim/wrapper callback that only executes `Waker::wake_by_ref` on the current waker associated with the future. The async runtime will then call `Future::poll` again which would then call the **user provided callback**.

- `wgpu_device_poll` should be called externally (one way or another, be it on another thread or something cranking in the event loop). With this setup it wouldn't matter if `wgpu_device_poll` was called from this thread or another thread as it's only triggering the waker, not actually calling the user callback. 


This _might_ even allow the lifetime constraint on the callback to be relaxed.

Reply:
I might be misunderstanding, but the user doesn't provide their own callbacks to wgpu-rs to be called by `wgpu_device_poll`. This is our own callback (internal to wgpu-rs) that we're passing to wgpu-native, so we could easily change it to call `wake_by_ref` if we'd prefer for it to do that. The user just sees futures for the mapped buffers.

(`wgpu_device_poll` will probably be called externally regardless, but ideally we'll add some kind of wgpu event loop to do that)

Reply:
> This is our own callback (internal to wgpu-rs) that we're passing to wgpu-native, so we could easily change it to call wake_by_ref if we'd prefer for it to do that. 

Yes, this is what I'm saying.

Reply:
@aloucks Any particular reason why using a waker internally is better than a callback?

Reply:
As I said before:
> With this setup it wouldn't matter if wgpu_device_poll was called from this thread or another thread as it's only triggering the waker, not actually calling the user callback.

And this (maybe):
> This _might_ even allow the lifetime constraint on the callback to be relaxed.

The callback would be executed wherever you have issued `.await` on the future. Where as if you run the user callback _directly_ from `device_poll` it could be executed from wherever `device_poll` was called. 

It's also how rust Futures are designed in general. The `Waker` exists so to notify the the runtime that the task is ready to do more work (i.e. polled again). 

Consider how async socket IO works. When you make a socket "asynchronous", somewhere in the bowels of tokio/mio/node.js/etc.., that socket is registered with `epoll`, `iocp`, `kqueue`, etc. Also deep in the bowels, there's a thread blocking on [epoll_wait](http://man7.org/linux/man-pages/man7/epoll.7.html). `epoll_wait` returns the event data that the caller can use to determine which sockets have data ready for read/write. 

The "IO driver" example that I pasted below was trying to emulate how this mechanism might work when there is no "real" IO. In our case, the notification system will have to be triggered by a GPU synchronization primitive (e.g. a fence or timeline semaphore) or something more course like the triggering from queue submission - or simply spinning in a loop with some kind of artificial throttle/delay. Having the "device poller" be a future itself that's spawned into the runtime may work too (I think this is what you have in the examples).

Reply:
So, replacing the internal implementation with `Waker` is very doable I think, but that'd still give us the option to fire callbacks from where device_poll is called, or in response to the future. Interestingly, I think firing the future without a special callback waker would be more efficient in Rust's case.